fiftyone/utils/ultralytics.py

```diff
@@ -270,3 +289,105 @@ def _to_detections(self, df):
                 for row in df.itertuples()
             ]
         )
+
+
+class FiftyOneYOLOModel(Model):
```

Comment:
Strictly speaking, the `fom.Model` interface requires that the `config` parameter to be an instance of a subclass of `fiftyone.core.config.Config` whose name is `<ModelClass>Config`.

So it would look more like this:

```py
from fiftyone.core.config import Config
from fiftyone.core.models import Model

class YOLOConfig(Config):
    def __init__(self, d):
        self.model = self.parse_raw(d, "model", default=None)

class YOLO(Model):
    def __init__(self, config):
        self.config = config
        self.model = config.model

config = YOLOConfig({"model": model})
model = YOLO(config)
```

Why do it like this when your current implementation works as written for your use case? The idea in general is that the `Config` class is supposed to be a serializable (ie something that can be expressed as JSON via `config.serialize()`) so that, for example, everything necessary to reload the model in another process can be stored in a file like [manifest-torch.json](https://github.com/voxel51/fiftyone/blob/develop/fiftyone/zoo/models/manifest-torch.json).

However, in this use case, the parameter is a preloaded `model` instance, so in the code above `config.serialize()` would not actually yield something that can be written to a JSON file.

Reply:
tldr, I'm mostly just taking the opportunity to explain the design behind the `fom.Model` interface's "normal" use case.

In this case, I think it's fine to keep `config` as a dict if you want for now because these models are not yet usable in JSON form in either approach.

Reply:
Totally understand. I would love to keep things consistent, especially if we are potentially going to add more integrations like this moving forward. What do you think about this approach:

It looks like YOLO models have a `ckpt_path` property, and this can be used to instantiate the model in the future, with `YOLO(checkpoint_path)` — this works for me. 

We don't want the model to be reloaded from disk just for the sake of serializability, but one option might be to allow the `YOLOConfig` to accept either a `YOLO` model, *or* a `checkpoint_path`, and then override the config's `serialize()` method to only pass along the `checkpoint_path`, after it is converted to absolute filepath. 

It might look something like this:

```
from fiftyone.core.config import Config
from fiftyone.core.models import Model

class YOLOConfig(Config):
    def __init__(self, d):
        self.checkpoint_path = self.parse_raw(d, "checkpoint_path", default=None)
        self.model = self.parse_raw(d, "model", default=None)
        
        if self.model and not self.checkpoint_path:
                 self.checkpoint_path = self.model.ckpt_path
        elif self.checkpoint_path and not self.model:
                self.model = YOLO(self.checkpoint_path)

        self.checkpoint_path = os.path.abspath(self.checkpoint_path)

   def serialize(self, reflective=False):
        """Serializes the object into a dictionary.

        Serialization is applied recursively to all attributes in the object,
        including element-wise serialization of lists and dictionary values.

        Args:
            reflective: whether to include reflective attributes when
                serializing the object. By default, this is False

        Returns:
            a JSON dictionary representation of the object
        """
        d = self._prepare_serial_dict(reflective)
        for a in self.attributes():
             if a == "model":
                continue
            d[a] = _recurse(getattr(self, a), reflective)
        return d

```

It seems like the serializability is encoded here: https://github.com/voxel51/eta/blob/ceb853415d7c468b553b5d49bbcffc1fb6216921/eta/core/serial.py#L406C4-L406C43

Reply:
Nice, `checkpoint_path` would be a great alternative to support!

WRT serialization, I was a bit strong with my wording: the most important thing is that the model can be *loaded* from JSON, at least in one possible configuration (`checkpoint_path` fulfills this here, so it's a 💯 addition); it's not actually necessary that all possible configurations actually be _serializable_. 

For example, [`TorchImageModelConfig`](https://github.com/voxel51/fiftyone/blob/acf3a8f886505d852903e320d057057813261993/fiftyone/utils/torch.py#L364) uses the same strategy of providing an optional `model` parameter where you can directly provide a `torch.nn.Module` in the case where you are constructing a Torch model via code. But I didn't bother modifying `serialize()` to manually omit that parameter.

Here's how I'd recommend implementing this (inspired by [TorchImageModel](https://github.com/voxel51/fiftyone/blob/acf3a8f886505d852903e320d057057813261993/fiftyone/utils/torch.py#L771)):

```py
class YOLOConfig(Config):
    def __init__(self, d):
        self.model = self.parse_raw(d, "model", default=None)
        self.checkpoint_path = self.parse_string(d, "checkpoint_path", default=None)

class YOLO(Model):
    def __init__(self, config):
        self.config = config
        self._model = self._load_model(config)

    def _load_model(self, config):
        if config.model is not None:
            return config.model

        return YOLO(config.checkpoint_path)
```

The reason is that the `Config` is _intended_ to be a static record of the input parameters that constructed the model. Any logic about how the parameters may be used is best done within the `Model` class so that you could check `model.config` to see how the model was defined.

Doesn't necessarily have any impact on the behavior of the model, just explaining my idea behind the `Model` + `ModelConfig` design 🤓

Reply:
Changes implemented :)